aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs

```diff
@@ -304,19 +304,40 @@ static PRECOMPILED_APTOS_FRAMEWORK_V2: Lazy<PrecompiledFilesModules> = Lazy::new
         .map(|(string, num_addr)| format!("{}={}", string, num_addr))
         .collect();
 
-    let options = move_compiler_v2::Options {
-        sources: aptos_cached_packages::head_release_bundle()
+    // aptos-experimental can be using latest language features, while others cannot.
+    // So we need to compile it twice, once without aptos-experimental, and once with.
+
+    let (experimental_sources, mut prod_sources): (Vec<String>, Vec<String>) =
+        aptos_cached_packages::head_release_bundle()
             .files()
-            .unwrap(),
+            .unwrap()
+            .into_iter()
+            .partition(|f| f.contains("aptos-move/framework/aptos-experimental/sources"));
+    let mut options = move_compiler_v2::Options {
+        sources: prod_sources.clone(),
         dependencies: vec![],
-        named_address_mapping: named_address_mapping_strings,
+        named_address_mapping: named_address_mapping_strings.clone(),
         known_attributes: aptos_framework::extended_checks::get_all_attribute_names().clone(),
         language_version: None,
         ..move_compiler_v2::Options::default()
     };
 
-    let (_global_env, modules) = move_compiler_v2::run_move_compiler_to_stderr(options)
-        .expect("stdlib compilation succeeds");
+    let (_global_env, mut modules) = move_compiler_v2::run_move_compiler_to_stderr(options.clone())
+        .expect("framework compilation succeeds");
+
+    // have experimental_sources at the start, so we can take the prefix of the result
+    // (but we need to provide both, in case aptos-experimental depends on other modules)
+    let experimental_sources_len = experimental_sources.len();
+    let mut all_sources = experimental_sources;
+    all_sources.append(&mut prod_sources);
+    options.sources = all_sources;
+    options.language_version = Some(LanguageVersion::latest());
+
+    let (_global_env, mut experimental_modules) =
+        move_compiler_v2::run_move_compiler_to_stderr(options)
+            .expect("aptos-experimental compilation succeeds");
+    experimental_modules.truncate(experimental_sources_len);
```

Comment:
This logic seems to be assuming that each source file can result in exactly one compiled module. However, one can have multiple modules in the same source file, resulting in that many compiled modules.

Thus, we could end up causing very subtle but dangerous bugs with this.

Reply:
hmmm. do you have a suggestion on how to fix this @vineethk ? 

looking at what run_move_compiler_to_stderr returns, there is package_name(&self) -> Option<Symbol>, but that is always None, and I don't see other info to help. Any ideas?

Reply:
Perhaps:
1. there is a `loc()` method we can call on each element of the `Vec<AnnotatedCompiledUnit>` returned by `run_move_compiler_to_stderr`.
2. the `_global_env` returned has a mapping from `FileHash` to filenames. `loc()` contains the `FileHash`.

This information could be used to filter the modules based on filenames?